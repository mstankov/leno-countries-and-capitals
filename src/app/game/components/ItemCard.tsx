"use client";

import { Card, CardProps } from "@mui/material";
import { ReadonlyPropsWithChildren } from "src/utils/types";

export type ItemCardProps<T extends string = string> = Omit<
  CardProps,
  "disabled" | "onClick"
> & {
  value: T;
  onClick: (country: T) => void;
  selected?: boolean;
  disabled?: boolean;
};

function ItemCard<T extends string>({
  value,
  onClick,
  selected,
  disabled,
  ...rest
}: ReadonlyPropsWithChildren<ItemCardProps<T>>) {
  const onClickHandler = () => {
    if (disabled) return;

    onClick(value);
  };

  return (
    <Card
      {...rest}
      data-testid={`card-${value}`}
      onClick={onClickHandler}
      sx={{
        p: 3,
        width: 200,
        textAlign: "center",
        backgroundColor: selected
          ? (theme) => theme.palette.warning.main
          : "white",
        cursor: disabled ? "not-allowed" : "pointer",
        transition: "background-color 0.2s ease-in-out",
        "&:hover": {
          backgroundColor: selected
            ? (theme) => theme.palette.warning.main[200]
            : (theme) => theme.palette.warning.light,
        },
        boxShadow: (theme) => (selected ? theme.shadows[5] : theme.shadows[1]),
        ...rest?.sx,
      }}
    >
      {value}
    </Card>
  );
}

export default ItemCard;
