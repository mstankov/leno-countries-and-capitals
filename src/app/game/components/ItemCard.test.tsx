import { render, screen } from "@testing-library/react";
import ItemCard from "./ItemCard";
import userEvent from "@testing-library/user-event";

describe(ItemCard.name, () => {
  it("should render", () => {
    const { container } = render(
      <ItemCard value="Belgium" onClick={jest.fn()} />
    );

    expect(container).toMatchSnapshot();
  });

  it("should call callback on click with country name", async () => {
    userEvent.setup();

    const onClick = jest.fn();
    render(<ItemCard value="Bulgaria" onClick={onClick} />);

    await userEvent.click(await screen.findByTestId("card-Bulgaria"));

    expect(onClick).toHaveBeenCalledTimes(1);
    expect(onClick).toHaveBeenCalledWith("Bulgaria");
  });

  it("should not call callback on click if component is disabled", async () => {
    userEvent.setup();

    const onClick = jest.fn();
    render(<ItemCard value="Bulgaria" onClick={onClick} disabled />);

    await userEvent.click(await screen.findByTestId("card-Bulgaria"));

    expect(onClick).not.toHaveBeenCalled();
  });
});
