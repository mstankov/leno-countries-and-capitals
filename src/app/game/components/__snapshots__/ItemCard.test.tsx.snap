// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemCard should render 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-51ttc2-MuiPaper-root-MuiCard-root"
    data-testid="card-Belgium"
  >
    Belgium
  </div>
</div>
`;
